### PR TITLE
Feature select module by path

### PR DESCRIPTION
**Pull request by achoimet**
- feature to search by properties_path instead of module.name
- feature to add properties at a given properties_path
- feature to add properties for one module to all his properties_path

### CODE REVIEW

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy [hunk 1]
```diff
@@ -155,9 +155,6 @@ class Hesperides implements Serializable {
 
     def putModuleOnPlatform(Map args) { required(args, ['app', 'platform', 'moduleName', 'moduleVersion', 'isWorkingCopy', 'logicGroupPath'])
         def platformInfo = getPlatformInfo(args)
-        if (listSelect(list: platformInfo.modules, key: 'name', value: args.moduleName)) {
```
Comment (Lucas-C) [inline src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: Why this change ?

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy [hunk 2]
```diff
@@ -296,7 +293,7 @@ class Hesperides implements Serializable {
             def modulePropertyChanges = propertyUpdates[moduleName]
             log "---------\n# Module: $moduleName\n---------"
             log "---------${platformInfo.modules}---------"
-            if (moduleName.contains('#')) {  // il s'agit de properties d'instance
+            if (moduleName.count('#')==1) {  // il s'agit de properties d'instance
```
Comment (Lucas-C) [inline src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: It may be better to check first for the `properties_path:` prefix, and then to check if the `moduleName` contains this character : no need to count then

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy (hunk 1 above)
Comment (achoimet) [reply src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: Because you based your implementation on the fact that one module exist only once on a platform, but it's false, it can exist several times at differents paths (it's the subject of the pull request) (ex: PAO :D )

File: test/HesperidesIntegrationSpec.groovy [hunk 3]
```diff
@@ -295,7 +328,7 @@ class HesperidesIntegrationSpec extends Specification implements Helper {
             hesperides.getInstanceProperties(app: applicationName, platform: platformName, moduleName: moduleName, instance: 'TOTO') == null
     }
 
-    def "Can delete all instances"() {
+/*    def "Can delete all instances"() {
```
Comment (Lucas-C) [inline test/HesperidesIntegrationSpec.groovy]: Comment forgotten ?

File: test/HesperidesIntegrationSpec.groovy [hunk 4]
```diff
@@ -48,28 +48,50 @@ class HesperidesIntegrationSpec extends Specification implements Helper {
...
+                logicGroupPath: "#${LogicGroupNameTwo}#${subLogicGroup}")
+        hesperides.createInstance(app: applicationName, platform: platformName, moduleName: moduleName, instance: instanceName, path: "#${logicGroupName}#${subLogicGroup}")
+        hesperides.createInstance(app: applicationName, platform: platformName, moduleName: secondModuleName, instance: instanceNameTwo, path: "#${logicGroupName}#${subLogicGroup}")
+        hesperides.createInstance(app: applicationName, platform: platformName, moduleName: secondModuleName, instance: instanceNameThree, path: "#${LogicGroupNameTwo}#${subLogicGroup}")
     }
 
     def cleanupSpec() {
-        hesperides.deletePlatform(app:applicationName, platform: platformName)
-        hesperides.deleteModule(moduleName: moduleName, version: moduleVersion, moduleType: 'workingcopy')
+        //hesperides.deletePlatform(app:applicationName, platform: platformName)
```
Comment (Lucas-C) [inline test/HesperidesIntegrationSpec.groovy]: Comments forgotten

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy [hunk 5]
```diff
@@ -548,14 +564,25 @@ class Hesperides implements Serializable {
     }
 
     @NonCPS
-    protected selectModule(Map args) { required(args, ['modules', 'moduleName'])
-        def module = args.modules.find { it.name == args.moduleName }
+    protected selectModule(Map args) { required(args, ['modules'])
+        def module
+        if (args.path && args.moduleName){
+            module = args.modules.find { it.name == args.moduleName && it.path == args.path}
+            log ("module trouvé : "+module.name)
```
Comment (Lucas-C) [inline src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: English: "trouvé" -> "found" ;)

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy [hunk 6]
```diff
@@ -548,14 +564,25 @@ class Hesperides implements Serializable {
     }
 
     @NonCPS
-    protected selectModule(Map args) { required(args, ['modules', 'moduleName'])
-        def module = args.modules.find { it.name == args.moduleName }
+    protected selectModule(Map args) { required(args, ['modules'])
+        def module
+        if (args.path && args.moduleName){
```
Comment (Lucas-C) [inline src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: Does this function takes an optional `path` or a `properties_path` parameter ? I see both used

**Comment by Lucas-C** [comment]
Could you please also update the doc with the new parameters you added in this PR ?
-> https://github.com/voyages-sncf-technologies/hesperides-jenkins-lib/blob/master/vars/hesperides.txt

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy (hunk 1 above)
Comment (Lucas-C) [reply src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: Alright ! Well done ;)

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy [hunk 7]
```diff
@@ -305,55 +302,70 @@ class Hesperides implements Serializable {
...
+                def modulePropertiesPath = moduleName.replace("properties_path:", "")
+                log "-> properties_path: $modulePropertiesPath"
+                def moduleFromPropertiesPath = selectModule(modules: platformInfo.modules, properties_path: modulePropertiesPath)
+                for (def y = 0; y < moduleFromPropertiesPath.instances.size(); y++){
+                    def instance = moduleFromPropertiesPath.instances[y].name
+                    def instanceInfo = extractInstanceInfo(modules: platformInfo.modules,
+                            properties_path: modulePropertiesPath,
+                            instance: instance)
+                    applyChanges(modulePropertyChanges, instanceInfo.key_values, "[instance=$instance] ")
+                    updatePlatform(platformInfo: platformInfo)
```
Comment (Lucas-C) [inline src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: Could this call be performed after the loop ?
This way we would make lesst HTTP calls

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy [hunk 8]
```diff
@@ -417,15 +429,19 @@ class Hesperides implements Serializable {
 
     @NonCPS
     private findModulePropertiesPath(Map args) { required(args, ['modules', 'moduleName']) // optional: instance
```
Comment (Lucas-C) [inline src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: 👍

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy [hunk 9]
```diff
@@ -305,55 +302,70 @@ class Hesperides implements Serializable {
...
-                    }
-                    def iterableValorisationItems = []
-                    for (def k = 0; k < newIterableItemProperties.size(); k++) { // DAMN Jenkins pipelines that does not support .eachWithIndex
-                        iterableValorisationItems << [title: 'not used', values: map2list(newIterableItemProperties[k], 'name', 'value')]
-                        // Displaying props changes:
-                        if (!actualIterableProperties) {
-                            continue
+                for (int p = 0; p < modulePropertiesPath.size(); p++){
+                    log("-> properties_path: "+modulePropertiesPath[p])
+                    log "passe la"
```
Comment (Lucas-C) [inline src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: Debug trace ;)

File: test/HesperidesIntegrationSpec.groovy [hunk 10]
```diff
@@ -277,7 +310,7 @@ class HesperidesIntegrationSpec extends Specification implements Helper {
         when:
             hesperides.createTemplate(moduleName: moduleName, moduleVersion: moduleVersion, location: '/etc', filename: 'titi', content: 'iam=titi')
             hesperides.createTemplate(moduleName: moduleName, moduleVersion: moduleVersion, location: '/etc', filename: 'toto', content: 'iam=toto')
-            def files = hesperides.getInstanceFiles(app: applicationName, platform: platformName, moduleName: moduleName, instance: instanceName)
+            def files = hesperides.getInstanceFiles(app: applicationName, platform: platformName, moduleName: moduleName, instance: instanceName, path: "#${logicGroupName}")
```
Comment (Lucas-C) [inline test/HesperidesIntegrationSpec.groovy]: Why this ?

File: src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy (hunk 6 above)
Comment (achoimet) [reply src/com/vsct/dt/hesperides/jenkins/pipelines/Hesperides.groovy]: I lost myself with path and properties_path... you're totally right, i will remove properties_path use (longer and complex compared to path property on hesperides) and use path only.

**Comment by achoimet** [comment]
All good !

**Comment by Lucas-C** [comment]
Awesome, well done 🎉🎉🎉
